qmxgraph/widget.py

```diff
@@ -158,22 +190,107 @@ def paintEvent(self, paint_event):
         p = QPainter(self)
         self.style().drawPrimitive(QStyle.PE_Widget, opt, p, self)
 
-    def load(self):
+    def call_once_when_loaded(self, function: Callable[[], None]) -> None:
         """
-        Load graph drawing page, if not yet loaded.
+        Calls the given function when the graph page has finished loading successfully,
+        or calls it immediately if the graph page is already loaded.
+        """
+        # TODO[ASIM-4285]: add tests.
+        if self.is_loaded():
+            function()
+        else:
+            self._call_once_loaded_callback.Register(function)
+
+    def _trigger_call_once_loaded_callback(self) -> None:
+        if self._web_view.view_state == ViewState.GraphLoaded:
+            self._call_once_loaded_callback()
+            self._call_once_loaded_callback.UnregisterAll()
+
+    def load_and_wait(self, *, timeout_ms: int=60_000) -> None:
+        """
+        Loads the graph page if not loaded yet, and blocks until it has been fully loaded.
+
+        Noop if the page is already loaded.
         """
-        if not self.is_loaded() or not self._web_view.is_loading():
-            self._load_graph_page()
+        # TODO[ASIM-4282]: add tests for this method.
+        # TODO[ASIM-4282]: rename to something more sensible, perhaps "ensure_loaded".
+        # TODO[ASIM-4282]: check if we can reduce that timeout.
+        if self._web_view.view_state == ViewState.GraphLoaded:
+            return
+        elif self._web_view.view_state == ViewState.Closing:
+            raise RuntimeError("view is closing, cannot load")
+
+        self.show()
+
+        if not self.is_loaded():
+            self.load()
+        wait_until(self.is_loaded, timeout_ms=timeout_ms, error_callback=lambda: f"view_state = {self._web_view.view_state}", wait_interval_ms=50)
+        return
+
+        def check_ok(*args) -> bool:
+            return self._web_view.view_state == ViewState.GraphLoaded
+
+        try:
+            #with wait_signals(self.loadFinished, timeout=timeout_ms, check_params_cb=check_ok):
```

Comment:
Commented code

Reply:
Tthis whole block is dead, there's a `return` a few lines above it, removed all the dead code. 👍